lib/util/stream_utils.js

```diff
@@ -30,55 +30,109 @@ shaka.util.StreamUtils = class {
+    const variants = StreamUtils.filterVariantsByAudioChannelCount(
         manifest.variants, preferredAudioChannelCount);
 
-    /**
-     * @param {!shaka.extern.Variant} variant
-     * @return {string}
-     */
-    function variantCodecs(variant) {
-      // Only consider the base of the codec string.  For example, these should
-      // both be considered the same codec: avc1.42c01e, avc1.4d401f
-      let baseVideoCodec = '';
-      if (variant.video) {
-        baseVideoCodec = MimeUtils.getCodecBase(variant.video.codecs);
-      }
+    // Now organize variants into buckets by codecs.
+    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+    let variantsByCodecs = StreamUtils.getVariantsByCodecs_(variants);
+    variantsByCodecs = StreamUtils.filterVariantsBySize_(variantsByCodecs);
 
-      let baseAudioCodec = '';
-      if (variant.audio) {
-        baseAudioCodec = MimeUtils.getCodecBase(variant.audio.codecs);
+    const bestCodecs = StreamUtils.findBestCodecs_(variantsByCodecs);
+
+    // Filter out any variants that don't match, forcing AbrManager to choose
+    // from the most efficient variants possible.
+    manifest.variants = manifest.variants.filter((variant) => {
+      const codecs = StreamUtils.getGroupVariantCodecs_(variant);
+      if (codecs == bestCodecs) {
+        return true;
       }
 
-      return baseVideoCodec + '-' + baseAudioCodec;
-    }
+      shaka.log.debug('Dropping Variant (better codec available)', variant);
+      return false;
+    });
+  }
 
-    // Now organize variants into buckets by codecs.
-    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+  /**
+  * Get variants by codecs.
+  *
+  * @param {!Array<shaka.extern.Variant>} variants
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  * @private
+  */
+  static getVariantsByCodecs_(variants) {
     const variantsByCodecs = new shaka.util.MultiMap();
     for (const variant of variants) {
-      const group = variantCodecs(variant);
+      const group = shaka.util.StreamUtils.getGroupVariantCodecs_(variant);
       variantsByCodecs.push(group, variant);
     }
 
-    // Compute the average bandwidth for each group of variants.
-    // Choose the lowest-bandwidth codecs.
-    let bestCodecs = null;
+    return variantsByCodecs;
+  }
+
+  /**
+  * Filters variants with missing sizes by all codecs.
+  *
+  * @param {!shaka.util.MultiMap.<shaka.extern.Variant>} variantsByCodecs
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  * @private
+  */
+  static filterVariantsBySize_(variantsByCodecs) {
+    let maxSize = 0;
+    const sizeByCodecs = new Map();
+    const countCodecs = variantsByCodecs.size();
+    variantsByCodecs.forEach((codecs, variants) => {
+      for (const variant of variants) {
+        const video = variant.video;
+        if (!video || !video.width || !video.height) {
+          continue;
+        }
+
+        const size = video.width * video.height + (video.frameRate || 0);
```

Comment:
There should be a number here. We are looking for the maximum resolution of all codecs.

Reply:
What's strange to me about this is that framerate is effectively "worth" about 1 pixel.  If you think of the "size" here (which is still a poor name) as pixels per second, you would multiply by the framerate (and default to 1) instead of adding it.

Reply:
Perhaps we should call it density. The frame rate is added to avoid collisions. I can do it with multiplication as well.